lazy-pages/src/host_func.rs

```diff
@@ -0,0 +1,238 @@
+// This file is part of Gear.
+
+// Copyright (C) 2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Host function call `pre_process_memory_accesses` support in lazy-pages.
+
+use std::{cell::RefMut, collections::BTreeSet};
+
+use gear_backend_common::{lazy_pages::Status, memory::ProcessAccessError};
+use gear_core::{
+    gas::GasLeft,
+    memory::{GranularityPage, MemoryInterval, PageU32Size, PagesIterInclusive},
+};
+
+use crate::{
+    common::{Error, LazyPage, LazyPagesExecutionContext},
+    process::{self, AccessHandler},
+    utils::{self, handle_psg_case_one_page},
+    LAZY_PAGES_PROGRAM_CONTEXT,
+};
+
+pub(crate) struct HostFuncAccessHandler<'a> {
+    pub is_write: bool,
+    pub gas_left: &'a mut GasLeft,
+}
+
+impl<'a> HostFuncAccessHandler<'a> {
+    fn sub_gas(&mut self, amount: u64) -> Status {
+        log::trace!("charge amount = {amount}");
+        self.gas_left.gas = if let Some(gas) = self.gas_left.gas.checked_sub(amount) {
+            gas
+        } else {
+            return Status::GasLimitExceeded;
+        };
+        self.gas_left.allowance = if let Some(gas) = self.gas_left.allowance.checked_sub(amount) {
+            gas
+        } else {
+            return Status::GasAllowanceExceeded;
+        };
```

Comment:
`let ... else` construction should be better for both cases

Reply:
Sorry but cannot agree with you)) `if let Some else` is really disappointed construction for me.